#!/bin/bash

# Input files:
# ftable.2.tsv : table of gene coordinates in the contigs of all assemblies analyzed
# p2g.2.tsv    : table to rename genes during construction of orthologous groups

# Concatenating Saccharum and Sorghum coding sequences
# Sorghum bicolor      : Sbicolor_313_v3.1.cds.fa
# Saccharum SP803280   : sc.mlc.cns.sgl.utg.scga7.CDS.codingseq.fasta
# Saccharum R570       : single_tiling_path_cds.fna
# Saccharum spontaneum : Sspon.v20171023.cds.fasta
cat Sbicolor_313_v3.1.cds.fa sc.mlc.cns.sgl.utg.scga7.CDS.codingseq.fasta single_tiling_path_cds.fna Sspon.v20171023.cds.fasta \
 > sacch_r570.cds.fa

# Format BLAST database and search all vs. all
makeblastdb -in sacch_r570.cds.fa -dbtype nucl -out sacch_r570.cds.db
blastn -db sacch_r570.cds.db -query sacch_r570.cds.fa -out sacch_r570.cds.blast.e-5.out \
 -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qlen slen nident" \
 -num_threads 10 -evalue 1e-5 -max_target_seqs 600000

# Calculate sequence length using EMBOSS's infoseq
infoseq -noheading -nocolumns -delimiter "\t" -only -name -length -outfile stdout -sequence sacch_r570.cds.fa > sacch_r570.cds.len

# Parse BLAST and build orthologous groups
prefix=sacch_r570.cds.blast.e-5
hsp2orthologs -vvv -r p2g.2.tsv -mio 10 -mao 500 -po 0.1 -s ${prefix}.tsv -of ${prefix}.og.tsv ${prefix}.out > ${prefix}.h2o.out 2> ${prefix}.h2o.err

# To remove the effect of fragmentation of the Saccharum SP803280 assembly,
# simulated contigs are generated by sampling fragments of the genomes of
# S. spontaneum and Saccharum R570 following the distribution of genes per
# contig observed in SP803280
seq 0 50 950 | parallel -n2 python simulation_r570_29082018.py {1} {2}
seq 50 50 1000 | parallel -n2 python simulation_r570_29082018.py {1} {2}
seq 0 50 950 | parallel -n2 python simulation_Sspon.py {1} {2}
seq 50 50 1000 | parallel -n2 python simulation_Sspon.py {1} {2}

# Concatenating simulation results
head -n1 simulation_novo100 > simulation.tsv
grep -v ^contig simulation_novo[0-9]* | cut -f 2 -d : >> simulation.tsv
grep Sspon simulation_novo_Sspon* | cut -f 2 -d : >> simulation.tsv

# After resampling, build a histogram (profile.svg) comparing the 
# distributions of the numbers of nearly syntenic blocks, shared by
# Sorghum and each Saccharum genomes, and a table describing such 
# blocks.
python3 figure.py
python3 projection.py
